[qt5-base] Add option to link to OpenSSL at compile-time

davidebeatrici opened the pull request:
By default Qt loads OpenSSL at runtime.

This pull request adds a `QT_OPENSSL_LINK` option, which specifies `-openssl-linked`. The option is enabled by default for static configurations.

From https://doc.qt.io/qt-5/ssl.html:

> By default, an SSL-enabled Qt library dynamically loads any installed OpenSSL library at run-time. However, it is possible to link against the library at compile-time by configuring Qt with the **-openssl-linked** option.

---

Fixes #2343.

Neumann-A commented:
make it a feature instead of changing the default behavior

davidebeatrici commented:
I agree, however the option should be enabled by default for static configurations.

Neumann-A commented:
>I agree, however the option should be enabled by default for static configurations.

only if OpenSSL has also been build statically. In custom triplets there is a possibility that OpenSSL is build dynamically and qt is build statically. So we need to figure out if OpenSSL has been built dynamically or not (e.g. by looking if there is a .dll or .so with the correct name).  You probably also need to pass `OPENSSL_LIBS` to make qt pass those libraries correctly in static builds.

davidebeatrici commented:
> only if OpenSSL has also been build statically. In custom triplets there is a possibility that OpenSSL is build dynamically and qt is build statically. So we need to figure out if OpenSSL has been built dynamically or not (e.g. by looking if there is a .dll or .so with the correct name).

Oh, I see.

We could use [FindOpenSSL](https://cmake.org/cmake/help/latest/module/FindOpenSSL.html)'s `OPENSSL_USE_STATIC_LIBS` option to check whether static libraries are available.

> You probably also need to pass `OPENSSL_LIBS` to make qt pass those libraries correctly in static builds.

I built `qt5-base` with this pull request's changes applied, I built a test application and there was no need to link to OpenSSL manually when linking to Qt.

JackBoosY commented:
/azp run

ras0219-msft commented:
If I understand the situation correctly, this is actually generally desirable regardless of whether openssl has been built statically or dynamically -- in the .so/.dll case, this will generate a proper loader directive from the Qt binary indicating its dependence on openssl instead of using `LoadLibrary()`/`dlopen()`.

Assuming I understand the situation, vcpkg will generally function better if this was the default behavior and users could disable it if they wanted via a triplet setting (not a feature).

ras0219-msft commented on ports/qt5-base/portfile.cmake:
```diff
@@ -1,5 +1,11 @@
 vcpkg_buildpath_length_warning(37)
 
+if(${VCPKG_LIBRARY_LINKAGE} STREQUAL "static")
```
A raw `${VAR}` is undesirable inside `if()` directives because it can result in double expansion -- this should instead be `if(VCPKG_LIBRARY_LINKAGE ...` or `if("${VCPKG_LIBRARY_LINKAGE}" ...` (both do the same thing).

davidebeatrici replied:
I followed the style that was used in the file, we should probably change all instances in a dedicated pull request.

davidebeatrici commented:
> If I understand the situation correctly, this is actually generally desirable regardless of whether openssl has been built statically or dynamically -- in the .so/.dll case, this will generate a proper loader directive from the Qt binary indicating its dependence on openssl instead of using `LoadLibrary()`/`dlopen()`.
>
> Assuming the above is correct, vcpkg will generally function better if this was the default behavior and users could disable it if they wanted via a triplet setting (not a feature).
>
> (Note: using `option()` does effectively implement a triplet setting)

I agree.

JackBoosY commented:
Please modify _ports/qt5-base/CONTROL_ and set `Version` to `5.12.5-3`.

JackBoosY reviewed (changes requested):
Please modify _ports/qt5-base/CONTROL_ and set `Version` to `5.12.5-3`.

davidebeatrici commented:
Sorry, I set the version to `5.12.5-3` and then reverted it to `5.12.5-2`: I accidentally thought I was the one who bumped the version to `5.12.5-2` while looking at the force-push changes (I rebased against `master`).

davidebeatrici committed:
[qt5-base] Add option to link to OpenSSL at compile-time

By default Qt loads OpenSSL at runtime.

This commit adds a "QT_OPENSSL_LINK" option, which specifies "-openssl-linked". The option is enabled by default for static configurations.

From https://doc.qt.io/qt-5/ssl.html:

"By default, an SSL-enabled Qt library dynamically loads any installed OpenSSL library at run-time. However, it is possible to link against the library at compile-time by configuring Qt with the -openssl-linked option."

ras0219-msft commented:
LGTM, thanks for the PR!

davidebeatrici commented:
You're welcome!
